ibis/backends/dask/core.py

```diff
@@ -0,0 +1,544 @@
+    ]
+
+    # if we're unable to find data then raise an exception
+    if not scopes and computable_args:
+        raise com.UnboundExpressionError(
+            'Unable to find data for expression:\n{}'.format(repr(expr))
+        )
+
+    # there should be exactly one dictionary per computable argument
+    assert len(computable_args) == len(scopes)
+
+    new_scope = new_scope.merge_scopes(scopes)
+    # pass our computed arguments to this node's execute_node implementation
+    data = [
+        new_scope.get_value(arg.op(), timecontext)
+        if hasattr(arg, 'op')
+        else arg
+        for arg in computable_args
+    ]
+    result = execute_node(
+        op,
+        *data,
+        scope=scope,
+        timecontext=timecontext,
+        aggcontext=aggcontext,
+        clients=clients,
+        **kwargs,
+    )
+    computed = post_execute_(op, result, timecontext=timecontext)
+    return Scope({op: computed}, timecontext)
+
+
+execute = Dispatcher('execute')
+
+
+@execute.register(ir.Expr)
+@trace
+def main_execute(
+    expr,
+    params=None,
+    scope=None,
+    timecontext: Optional[TimeContext] = None,
+    aggcontext=None,
+    **kwargs,
+):
+    """Execute an expression against data that are bound to it. If no data
+    are bound, raise an Exception.
+
+    Parameters
+    ----------
+    expr : ibis.expr.types.Expr
+        The expression to execute
+    params : Mapping[ibis.expr.types.Expr, object]
+        The data that an unbound parameter in `expr` maps to
+    scope : Mapping[ibis.expr.operations.Node, object]
+        Additional scope, mapping ibis operations to data
+    timecontext : Optional[TimeContext]
+        timecontext needed for execution
+    aggcontext : Optional[ibis.dask.aggcontext.AggregationContext]
+        An object indicating how to compute aggregations. For example,
+        a rolling mean needs to be computed differently than the mean of a
+        column.
+    kwargs : Dict[str, object]
+        Additional arguments that can potentially be used by individual node
+        execution
+
+    Returns
+    -------
+    result : Union[
+        dask.dataframe.Series,
+        dask.dataframe.DataFrame,
+        ibis.dask.core.simple_types
+    ]
+
+    Raises
+    ------
+    ValueError
+        * If no data are bound to the input expression
+    """
+
+    if scope is None:
+        scope = Scope()
+
+    if timecontext is not None:
+        # convert timecontext to datetime type, if time strings are provided
+        timecontext = canonicalize_context(timecontext)
+
+    if params is None:
+        params = {}
+
+    # TODO: make expresions hashable so that we can get rid of these .op()
+    # calls everywhere
+    params = {k.op() if hasattr(k, 'op') else k: v for k, v in params.items()}
+    scope = scope.merge_scope(Scope(params, timecontext))
+    return execute_with_scope(
+        expr, scope, timecontext=timecontext, aggcontext=aggcontext, **kwargs,
+    )
+
+
+def execute_and_reset(
```

Comment:
Does this method actually trigger the dask job? Or does it just return a deferred dask.DataFrame that is executed later?

Reply:
As much as possible the dask backend does not run `compute` and returns a dask expression that the user can choose to execute. Sometimes it is unavoidable, but that's the goal.

Reply:
kk I understand this now - "execute" does not actually trigger computation, but return a dd.DataFrame/Series instead.